### ISSUE TITLE
TypeError: Cannot read property 'length' of null

### ISSUE THREAD

**Issue by agilenut** [rfennell/AzurePipelines#822]
**Description**
Task: Generate Release Notes (Crossplatform) V3
Location: Release
Settings:
- Output File: $(System.DefaultWorkingDirectory)\releasenotes.md
- Template Location: Inline
- Template: Test text
- Replace Output File: True

**To Reproduce**
1. Run task with any template
2. See error in output

**Expected behavior**
Successfully generates file

**Screenshots**
![image](https://user-images.githubusercontent.com/705864/94611104-3fcda800-026f-11eb-9237-681835a662ea.png)

**Additional context**
* On Prem Azure Dev Ops - how can I get version info from UI?
* Doesn't matter what content is used in template.
* Seems to fail while gathering WI info.
* We do not use WI's in TFS.
* Tested in Azure Dev Ops in cloud and it worked.

**Comment by rfennell** [comment]
Looks like there is a null check missing, the code is expecting an empty array when it get a null. I will add an extra check and push out a new version

**Comment by agilenut** [comment]
@rfennell,

Thank you for the quick update. 

I pulled the new version (3.16.2) but I am still getting the same "Cannot property 'length' of null" errors.

**Comment by rfennell** [comment]
That is strange, the only place the log message prior to the error now has an error trap on the next line. So it looks as if might be to do with next block on before the next log message. That code is all to do getting with commits. Those calls should all return at least an empty array, but I guess there must be a problem

I will see if I can get a repo with on an on prem. server.

**Comment by rfennell** [comment]
I can't repo the issue on my Azure DevOps 2019 server. In the log you should something along the lines next

> Detected 0 commits/changesets and 0 workitems between the current build and the last successful one
> Detected 0 tests associated within the current build.

The only reason it could fail before these lines is if the commit array is null and not an empty array.

So it is strange that this is only effecting you and on premises

To try to give a means to progress this, I have attached an updated .JS (in a Zip file) to try to progress this issue, could you drop this on your build agent by hand to get more detailed logs.

The steps are

1. The file needs to  replace the one of the same name in your agent's  `_work\_tasks\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\3.16.2`
1. Once this is done rerun your release and you see extra logging that show each step after the last good logged line. Note that have also added code to resets the `commits` and `workitems` objects to empty arrays if the object is null.

If this fixes the issue I will publish it in the normal manner

[ReleaseNotesFunctions.zip](https://github.com/rfennell/AzurePipelines/files/5306926/ReleaseNotesFunctions.zip)

**Comment by agilenut** [comment]
I will give it a try.

If it is any help, we have this under Project Settings -> Board -> Team Configurations: 

TF400917: The current configuration is not valid for this feature. This feature cannot be used until you correct the configuration. To attempt to fix the configuration, run Feature Enablement or correct your configuration manually.

I will ask about this today to see if anyone knows why this is happening. But I'm wondering if we don't have access to use WIs and the logic is failing to query that data.

**Comment by agilenut** [comment]
@rfennell,

I've tested the new JS file. The release does run successfully. However, the output does not include commits. I assume that is because of the code that resets commits to []. But obviously, I'd need the build's commits to include in the final version.

**Comment by rfennell** [comment]
Ok that is some progress, however the fixed JS file should still return commits if there are any. All the fix does is reset the `commits` object to an empty array if it is null (but it should never be a null).If it contains data it is left alone.

So the question is why does the API call (though called via the Node SDK we are really calling [this REST call](https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/get%20changes%20between%20builds?view=azure-devops-server-rest-5.0)) return a null, when I would expect an empty array at least.

The call for Azure DevOps Server 2019 is 
```
https://{instance}/{collection}/{project}/_apis/build/changes?fromBuildId={fromBuildId}&toBuildId={toBuildId}&api-version=5.0-preview.2
```

From your initial screenshot of the debug log you can see the `fromBuildID` should be 31366 and the `toBuildid` should be 31369. So in an authenticated browser, if you swap the instance, collection and project you can enter the following URL

```
https://{instance}/{collection}/{project}/_apis/build/changes?fromBuildId=31366&toBuildId=31369&api-version=5.0-preview.2
````
 and expect to see a JSON payload returned.

Can you try this?

Also what version of Azure DevOps Server are you running (you can find it on the Azure DevOps Management console), I am wondering if the issue is API version related?

**Comment by agilenut** [comment]
I can run the api call and I get a single changeset in the array which is correct. I ran it again for a different build pair and got the same. So, the API call seems to be working. 

I currently only have access to the browser UI (with limited permissions) and a single build agent. I'm asking someone else that has more permissions which version we have.

I also asked about the configuration error that I mentioned above in our standup. Apparently, since we don't use WI's we never bothered to configure them and would rather not since it won't be used. So, those features are not enabled.

**Comment by agilenut** [comment]
Here are what appears to be the pertinent logs from a run with the newer js file:

```
2020-09-30T17:50:16.9258789Z Adding the build [31416] and its association to the unified results object
2020-09-30T17:50:16.9262348Z Get details of [0] WIs
2020-09-30T17:50:16.9262635Z Complete getFullWorkItemDetails
2020-09-30T17:50:16.9266871Z Complete UnifiedArtifactDetails
2020-09-30T17:50:16.9267111Z commits are null setting as an empty array
2020-09-30T17:50:16.9269968Z Detected 0 commits/changesets and 0 workitems between the current build and the last successful one
2020-09-30T17:50:16.9270091Z Detected 0 tests associated within the current build.
2020-09-30T17:50:16.9271980Z 
2020-09-30T17:50:16.9436224Z ##[debug]No tests associated with release 95 environment Publish Release Notes
2020-09-30T17:50:16.9436559Z ##[debug]The global test array contains 0 items
2020-09-30T17:50:16.9436723Z ##[debug]The updated global test array contains 0 items
2020-09-30T17:50:16.9437360Z Removing duplicate Commits from master list
2020-09-30T17:50:16.9439125Z Removing duplicate WorkItems from master list
2020-09-30T17:50:16.9449291Z Expanding the truncated commit messages...
2020-09-30T17:50:16.9452271Z Expanded 0
2020-09-30T17:50:16.9453554Z Get details of [0] WIs
2020-09-30T17:50:16.9454499Z Total build artifacts: [1]
2020-09-30T17:50:16.9455204Z Total commits: [0]
2020-09-30T17:50:16.9455843Z Total workitems: [0]
2020-09-30T17:50:16.9456504Z Total related workitems: [0]
2020-09-30T17:50:16.9457154Z Total release tests: [0]
2020-09-30T17:50:16.9457804Z Total tests: [0]
2020-09-30T17:50:16.9458430Z Leaving WI in default order as returned by API
2020-09-30T17:50:16.9459224Z ##[debug]Getting the details of build 31416 from default project
2020-09-30T17:50:16.9618089Z The default artifact for the release was not linked to an Azure DevOps Git Repo Pull Request
2020-09-30T17:50:16.9618434Z Getting all completed Azure DevOps Git Repo PRs in the Team Project Sandbox
2020-09-30T17:50:16.9858395Z ##[debug]No completed Azure DevOps PRs found
2020-09-30T17:50:16.9860127Z Total Pull Requests: [0]
```

So, the logs show null commits but the api call does not.

**Comment by agilenut** [comment]
@rfennell,

I think I may have found the issue and I believe you are correct that it is an API problem. The current version of the azure-devops-node-api is using 5.1-preview.2 not 5.0-preview.2 as you had me test. 

The 5.0-preview2 works but the 5.1-preview2 version does not.

**Comment by rfennell** [comment]
I have been testing the task on Azure DevOps Server 2019.Update1.1 17.153.29522.3 and it seem to work Ok.

Just to match your setup I tested with TFVC (most of my testing is with it these days as I find there are few TFVC users of this task) and that worked too.

So can you find the version of Azure DevOps/TFS you are using, I suspect you are on an unpatched 2019. 

Problem is I don't intend to roll back the API version I target as that effects the feature I use. So my recommendation will be upgrade the server if it is not patched up to date. I always recommend patching/upgrading on premise TFS/Azure DevOps server every 6months at least to the latest features. Microsoft tend to ship 3 or 4 release a year, we are expecting Azure DevOps Server 2020 anytime soon.

**Comment by agilenut** [comment]
Thanks for confirming. 

I suspect you are correct. And, I wouldn't expect you to downgrade the node package or API version. 

I will look into getting the server patched. However, that might take some negotiating. 

I do have your PowerShell version working. If the server upgrade is not possible, I do I have 2 questions about that version:

1. Is it still being maintained now that the cross-platform one exists?
2. I'm using it with the Wiki Updater task and it seems I'm getting a line ending artifact when pre-pending pages which causes markdown headings to not display correctly. Is that a bug or is it expected that we move from the PowerShell version to the cross-platform version to fix it?

**Comment by rfennell** [comment]
In answer to your questions

1. The older Powershell version is not under active development by me. I am doing bug fixes and considering submitted PR if submitted but that is all. My plan was to monitor the usage and aim to deprecate it formally in a year or so. There are two reasons
    - It used my own homegrown logic work out WI/CS related to releases, whereas this Node version uses the official Microsoft API calls
    - It's templating language is far less flexible than Handlebars used by the Node version
1. In the WIKIUpdater in prepend mode it does not a line feed between the new and old block. Bit it does make sure the line feed format is corrent
```
             var oldContent = "";
                if (fs.existsSync(workingFile)) {
                    oldContent = fs.readFileSync(workingFile, "utf8");
                }
                fs.writeFileSync(workingFile, contents.replace(/`n/g, "\r\n"));
                fs.appendFileSync(workingFile, oldContent);
```

**Comment by agilenut** [comment]
Understood on the plans to deprecate.

Regarding the line ending, here is what I'm seeing (happy to move this to a new issue if needed)

Template:
```
## Version $($build.buildnumber)

@@CSLOOP@@
* $($csdetail.comment) ($($csdetail.changesetid)$($csdetail.commitid))
@@CSLOOP@@
```

Output:
![image](https://user-images.githubusercontent.com/705864/94832031-6b6da100-03db-11eb-9d81-b14b57cda103.png)

Where my cursor is, there seems to be a 0 width character between the new line and the first # sign. If I delete the character, I get this:
![image](https://user-images.githubusercontent.com/705864/94832121-82ac8e80-03db-11eb-82cf-1b99c8a3dce3.png)

I also noticed if I try copying the output into a file in Visual Studio and save it says there are characters that can't be saved in Unicode.

**Comment by rfennell** [comment]
It would be easier if that is separate issue if you don't mind, but certainly something I can look into.

Just to let you know, I am just about to head off on a long weekend, so I doubt I will be able to look into the WIKI issue until next week

**Comment by fbadaro** [comment]
I have the same problem in the release pipeline, in my case the agent is in the cloud, and the 'length' error happened here, in the same way, the task of updating the wiki with 'replace' problems in script.
